### ISSUE TITLE
Fail a whole pipe when a single command fails (equivalent to `set -o pipefail`)

### ISSUE THREAD

**Issue by sama-nv** [sindresorhus/execa#871]
Is there a way to configure execa to fail a whole pipe when a single command fails? The equivalent in shell is `set -o pipefail`.

e.g. I want this to throw an error, since the first stage fails.

```js
await execa('ls', ['--bad']).pipeStdout(execa('cat'));
```

I could prefix the piped command with `set -o pipefail;` with `shell: true`, but this not recommended and less secure:

```js
await execa('set -o pipefail; ls --bad | cat', { shell: true });
```

**Comment by ehmicky** [comment]
Hi @sama-nv,

You're in luck, we just added this very feature.

https://github.com/sindresorhus/execa/blob/c8ba4269fb8bd3e2681ac7953343766e81ad2304/lib/pipe/sequence.js#L2

The [new API](https://github.com/sindresorhus/execa?tab=readme-ov-file#pipefile-arguments-options) looks like this:

```js
await execa('ls', ['--bad']).pipe('cat');
```

```
Error: Command failed with exit code 2: ls --bad

ls: unrecognized option '--bad'
Try 'ls --help' for more information.
    at createFinalError (file:///.../execa/lib/return/clone.js:9:10)
    at ... {
  shortMessage: 'Command failed with exit code 2: ls --bad',
  originalMessage: '',
  command: 'ls --bad',
  escapedCommand: 'ls --bad',
  cwd: '/home/ether/code/execa',
  failed: true,
  timedOut: false,
  isCanceled: false,
  isTerminated: false,
  exitCode: 2,
  signal: undefined,
  signalDescription: undefined,
  stdout: '',
  stderr: "ls: unrecognized option '--bad'\nTry 'ls --help' for more information.",
  stdio: [
    undefined,
    '',
    "ls: unrecognized option '--bad'\nTry 'ls --help' for more information."
  ],
  pipedFrom: []
}
```

What happens step-by-step is:
  1. `ls --bad` completes with exit code 2
  2. As a result, the `stdout` from `ls` ends
  3. This stops the piping from `ls` `stdout` to `cat` `stdin`
  4. As a result, the `stdin` from `cat` ends
  5. With no more data to read, the `cat` process completes with exit code 0
  6. Execa notices that one of the two processes had a non-0 exit code. As a consequence, the `await`ed promise is rejected with the result of `execa('ls', ['--bad'])`.

This is actually the same behavior as in Bash with `pipefail`.

```sh
$ set -o pipefail

$ ls --bad | cat
ls: unrecognized option '--bad'
Try 'ls --help' for more information.

$ echo "$?" "${PIPESTATUS[*]}" 
2 2 0
```

Unfortunately, this has not been released yet. We have multiple big changes coming up for the next release, so it might take a few more weeks.

**Comment by ehmicky** [comment]
@sindresorhus It is nice to see we've been working on the right thing with piping processes! :arrow_up:

**Comment by sama-nv** [comment]
Thanks! This is exactly what I need. I can probably come up with a workaround in the meantime.

**Comment by ehmicky** [comment]
You can use the following workaround in the meantime.

```js
const source = execa('ls', ['--bad'])
const [result] = await Promise.all([source.pipeStdout(execa('cat')), source]);
```

**Comment by ehmicky** [comment]
This feature has been just released in Execa 9.0.0. Please see (and share!) the [release post](https://medium.com/@ehmicky/execa-9-release-d0d5daaa097f) and the [changelog](https://github.com/sindresorhus/execa/releases/tag/v9.0.0).
